File: CartQuestAndroid/app/src/main/java/com/amoghghadge/cartquestandroid/ui/route/RouteMapViewModel.kt
package com.amoghghadge.cartquestandroid.ui.route

import android.app.Application
import android.content.Intent
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.viewModelScope
import com.amoghghadge.cartquestandroid.BuildConfig
import com.amoghghadge.cartquestandroid.data.model.CompletedRun
import com.amoghghadge.cartquestandroid.data.remote.DirectionsApiService
import com.amoghghadge.cartquestandroid.data.remote.KrogerApiService
import com.amoghghadge.cartquestandroid.data.remote.KrogerAuthManager
import com.amoghghadge.cartquestandroid.data.repository.CartRepository
import com.amoghghadge.cartquestandroid.data.repository.RunsRepository
import com.amoghghadge.cartquestandroid.service.LocationService
import com.amoghghadge.cartquestandroid.service.RouteOptimizer
import com.google.android.gms.maps.model.LatLng
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

sealed class RouteState {
    object Loading : RouteState()
    data class Computed(val route: RouteOptimizer.OptimizedRoute, val userLocation: LatLng) : RouteState()
    data class Error(val message: String) : RouteState()
}

class RouteMapViewModel(
    application: Application,
    savedStateHandle: SavedStateHandle
) : AndroidViewModel(application) {

    private val cartId: String = savedStateHandle["cartId"] ?: ""
    private val cartRepository = CartRepository()
    private val runsRepository = RunsRepository()
    private val locationService = LocationService(application)
    private val krogerApi = KrogerApiService.create()
    private val krogerAuth = KrogerAuthManager(
        clientId = BuildConfig.KROGER_CLIENT_ID,
        clientSecret = BuildConfig.KROGER_CLIENT_SECRET
    )
    private val directionsApi = DirectionsApiService(BuildConfig.GOOGLE_MAPS_API_KEY)
    private val optimizer = RouteOptimizer()

    private val _routeState = MutableStateFlow<RouteState>(RouteState.Loading)
    val routeState: StateFlow<RouteState> = _routeState

    private val _tripCompleted = MutableStateFlow(false)
    val tripCompleted: StateFlow<Boolean> = _tripCompleted

    init {
        computeRoute()
    }

    fun computeRoute() {
        viewModelScope.launch {
            _routeState.value = RouteState.Loading
            try {
                // 1. Load cart
                val cart = cartRepository.getActiveCart()
                    ?: throw IllegalStateException("No active cart found")
                val cartItems = cart.items
                if (cartItems.isEmpty()) throw IllegalStateException("Cart is empty")

                // 2. Get user location
                val userLocation = locationService.getCurrentLocation()

                // 3. Query nearby stores
                val token = krogerAuth.getToken()
                val storesResponse = krogerApi.searchLocations(
                    auth = "Bearer $token",
                    lat = userLocation.latitude,
                    lon = userLocation.longitude
                )
                val stores = storesResponse.data
                if (stores.isEmpty()) throw IllegalStateException("No nearby stores found")

                // 4. For each store, query product availability (parallelized)
                val allProductIds = cartItems.flatMap { item ->
                    listOf(item.productId) + item.substitutes.map { it.productId }
                }.distinct()

                val storeAvailabilities = stores.map { store ->
                    async {
                        try {
                            // Query all products for this store in parallel
                            val productResults = allProductIds.map { pid ->
                                async {
                                    try {
                                        val response = krogerApi.searchProducts(
                                            auth = "Bearer $token",
                                            term = pid,
                                            locationId = store.locationId,
                                            limit = 1
                                        )
                                        val product = response.data.firstOrNull { it.productId == pid }
                                        if (product != null) pid to product else null
                                    } catch (_: Exception) {
                                        null
                                    }
                                }
                            }.awaitAll().filterNotNull()

                            if (productResults.isNotEmpty()) {
                                RouteOptimizer.StoreAvailability(store, productResults.toMap())
                            } else null
                        } catch (_: Exception) {
                            null
                        }
                    }
                }.awaitAll().filterNotNull()

                if (storeAvailabilities.isEmpty()) {
                    throw IllegalStateException("No stores have the requested products")
                }

                // 5. Run optimizer
                val optimizedRoute = optimizer.optimize(
                    cartItems = cartItems,
                    storeAvailabilities = storeAvailabilities,
                    userLocation = userLocation,
                    getDriveTime = { origin, storeList ->
                        val waypoints = storeList.map {
                            LatLng(it.geolocation.latitude, it.geolocation.longitude)
                        }
                        val destination = waypoints.last()
                        val intermediateWaypoints = if (waypoints.size > 1) waypoints.dropLast(1) else emptyList()
                        val result = directionsApi.getDirections(
                            origin = origin,
                            destination = destination,
                            waypoints = intermediateWaypoints
                        )
                        Pair(result.totalDurationSeconds, result.encodedPolyline)
                    }
                )

                _routeState.value = RouteState.Computed(optimizedRoute, userLocation)
            } catch (e: Exception) {
                _routeState.value = RouteState.Error(e.message ?: "Unknown error")
            }
        }
    }

    fun startNavigation(context: android.content.Context) {
        val state = _routeState.value
        if (state !is RouteState.Computed) return

        val stops = state.route.stops
        if (stops.isEmpty()) return

        val userLocation = state.userLocation

        // Build Google Maps dir URL with all stops in route order
        val pathStops = stops.joinToString("/") { "${it.lat},${it.lng}" }
        val uri = if (stops.size > 1) {
            Uri.parse("https://www.google.com/maps/dir/${userLocation.latitude},${userLocation.longitude}/$pathStops")
        } else {
            Uri.parse("google.navigation:q=${stops.first().lat},${stops.first().lng}")
        }

        val intent = Intent(Intent.ACTION_VIEW, uri).apply {
            setPackage("com.google.android.apps.maps")
        }
        if (intent.resolveActivity(context.packageManager) != null) {
            context.startActivity(intent)
        } else {
            // Fall back to browser if Google Maps not installed
            val browserIntent = Intent(Intent.ACTION_VIEW, uri)
            context.startActivity(browserIntent)
        }
    }

    fun completeTrip() {
        viewModelScope.launch {
            try {
                val state = _routeState.value
                if (state !is RouteState.Computed) return@launch

                val user = Firebase.auth.currentUser
                val run = CompletedRun(
                    userId = user?.uid ?: "",
                    displayName = user?.displayName ?: "",
                    photoUrl = user?.photoUrl?.toString() ?: "",
                    completedAt = System.currentTimeMillis(),
                    stores = state.route.stops,
                    totalDriveTimeMinutes = state.route.totalDriveTimeSeconds / 60,
                    totalCost = state.route.stops.sumOf { stop ->
                        stop.items.sumOf { it.price }
                    }
                )
                runsRepository.saveCompletedRun(run)
                cartRepository.completeCart(cartId)
                _tripCompleted.value = true
            } catch (_: Exception) {
                // handle silently for now
            }
        }
    }
}

File: CartQuestiOS/CartQuestiOS/Views/Route/RouteMapViewModel.swift
import Foundation
import UIKit
import CoreLocation
import FirebaseAuth
import Observation

enum RouteState: Equatable {
    case loading
    case computed(route: RouteOptimizer.OptimizedRoute, userLocation: CLLocationCoordinate2D)
    case error(String)

    static func == (lhs: RouteState, rhs: RouteState) -> Bool {
        switch (lhs, rhs) {
        case (.loading, .loading):
            return true
        case (.error(let a), .error(let b)):
            return a == b
        case (.computed(let r1, let l1), .computed(let r2, let l2)):
            return r1.encodedPolyline == r2.encodedPolyline
                && l1.latitude == l2.latitude
                && l1.longitude == l2.longitude
        default:
            return false
        }
    }
}

@Observable
class RouteMapViewModel {
    var routeState: RouteState = .loading
    var tripCompleted: Bool = false

    private let cartId: String
    private let cartRepository: CartRepository
    private let runsRepository: RunsRepository
    private let krogerService: KrogerService
    private let locationService: LocationService
    private let directionsService: DirectionsService
    private let routeOptimizer: RouteOptimizer

    private var cart: Cart?

    init(
        cartId: String,
        cartRepository: CartRepository = CartRepository(),
        runsRepository: RunsRepository = RunsRepository(),
        krogerService: KrogerService = KrogerService(
            clientId: "KROGER_CLIENT_ID_PLACEHOLDER",
            clientSecret: "KROGER_CLIENT_SECRET_PLACEHOLDER"
        ),
        locationService: LocationService = LocationService(),
        directionsService: DirectionsService = DirectionsService(apiKey: "GOOGLE_MAPS_API_KEY_PLACEHOLDER"),
        routeOptimizer: RouteOptimizer = RouteOptimizer()
    ) {
        self.cartId = cartId
        self.cartRepository = cartRepository
        self.runsRepository = runsRepository
        self.krogerService = krogerService
        self.locationService = locationService
        self.directionsService = directionsService
        self.routeOptimizer = routeOptimizer

        Task { await computeRoute() }
    }

    // MARK: - Route Computation

    private func computeRoute() async {
        routeState = .loading
        do {
            // 1. Load cart
            guard let activeCart = try await cartRepository.getActiveCart(),
                  !activeCart.items.isEmpty else {
                routeState = .error("Cart is empty or not found.")
                return
            }
            self.cart = activeCart

            // 2. Get user location
            let userLocation = try await locationService.getCurrentLocation()

            // 3. Query nearby stores
            let stores = try await krogerService.searchLocations(
                lat: userLocation.latitude,
                lon: userLocation.longitude,
                radiusInMiles: 10,
                limit: 10
            )

            guard !stores.isEmpty else {
                routeState = .error("No stores found nearby.")
                return
            }

            // 4. For each store, query product availability
            var storeAvailabilities: [RouteOptimizer.StoreAvailability] = []
            let allProductIds = activeCart.items.flatMap { item in
                [item.productId] + item.substitutes.map { $0.productId }
            }
            let uniqueProductIds = Array(Set(allProductIds))

            for store in stores {
                var availableProducts: [String: KrogerProduct] = [:]
                // Search for each product at this store location
                for productId in uniqueProductIds {
                    do {
                        let results = try await krogerService.searchProducts(
                            term: productId,
                            locationId: store.locationId,
                            limit: 1
                        )
                        if let product = results.first(where: { $0.productId == productId }) {
                            availableProducts[productId] = product
                        }
                    } catch {
                        // Skip product if query fails for this store
                        continue
                    }
                }
                if !availableProducts.isEmpty {
                    storeAvailabilities.append(
                        RouteOptimizer.StoreAvailability(
                            store: store,
                            availableProducts: availableProducts
                        )
                    )
                }
            }

            // 5 & 6. Run optimizer
            let route = try await routeOptimizer.optimize(
                cartItems: activeCart.items,
                storeAvailabilities: storeAvailabilities,
                userLocation: userLocation,
                getDriveTime: { [weak self] origin, storeList in
                    guard let self else { throw URLError(.cancelled) }
                    let waypoints = storeList.dropLast().map {
                        CLLocationCoordinate2D(latitude: $0.geolocation.latitude, longitude: $0.geolocation.longitude)
                    }
                    let lastStore = storeList.last!
                    let destination = CLLocationCoordinate2D(
                        latitude: lastStore.geolocation.latitude,
                        longitude: lastStore.geolocation.longitude
                    )
                    let result = try await self.directionsService.getDirections(
                        origin: origin,
                        destination: destination,
                        waypoints: Array(waypoints)
                    )
                    return (result.totalDurationSeconds, result.encodedPolyline)
                }
            )

            routeState = .computed(route: route, userLocation: userLocation)
        } catch {
            routeState = .error(error.localizedDescription)
        }
    }

    // MARK: - Navigation

    func startNavigation() {
        guard case .computed(let route, let userLocation) = routeState else { return }

        let stops = route.stops
        guard !stops.isEmpty else { return }

        // Build Google Maps URL with waypoints
        let destination = stops.last!
        let destinationStr = "\(destination.lat),\(destination.lng)"

        var urlString = "comgooglemaps://?saddr=\(userLocation.latitude),\(userLocation.longitude)"
        urlString += "&daddr=\(destinationStr)"

        if stops.count > 1 {
            let waypointStops = stops.dropLast()
            let waypointsStr = waypointStops.map { "\($0.lat),\($0.lng)" }.joined(separator: "+to:")
            urlString += "+to:\(waypointsStr)"
        }

        urlString += "&directionsmode=driving"

        if let url = URL(string: urlString) {
            UIApplication.shared.open(url)
        } else if let mapsUrl = URL(string: "https://www.google.com/maps/dir/?api=1&origin=\(userLocation.latitude),\(userLocation.longitude)&destination=\(destinationStr)&travelmode=driving") {
            UIApplication.shared.open(mapsUrl)
        }
    }

    // MARK: - Trip Completion

    func completeTrip() async {
        guard case .computed(let route, _) = routeState else { return }

        do {
            let user = Auth.auth().currentUser
            let totalCost = route.stops.flatMap { $0.items }.reduce(0.0) { $0 + $1.price }

            let run = CompletedRun(
                userId: user?.uid ?? "",
                displayName: user?.displayName ?? "Anonymous",
                photoUrl: user?.photoURL?.absoluteString ?? "",
                completedAt: Date(),
                stores: route.stops,
                totalDriveTimeMinutes: route.totalDriveTimeSeconds / 60,
                totalCost: totalCost
            )

            _ = try await runsRepository.saveCompletedRun(run)

            // Mark cart as completed
            if let cart = self.cart, !cart.id.isEmpty {
                try await cartRepository.completeCart(cartId: cart.id)
            }

            tripCompleted = true
        } catch {
            routeState = .error("Failed to save trip: \(error.localizedDescription)")
        }
    }
}
